core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java

```diff
@@ -818,6 +857,23 @@ private static AnnotatedValueResolver ofInjectableTypes0(String name, AnnotatedE
         };
     }
 
+    /**
+     * Returns an attribute resolver which retrieves a value specified by {@code attrKeys}
+     * from the {@link RequestContext}.
+     */
+    private static BiFunction<AnnotatedValueResolver, ResolverContext, Object>
+    attributeResolver(Iterable<AttributeKey<?>> attrKeys) {
+        return (resolver, ctx) -> {
+            for (AttributeKey<?> attrKey : attrKeys) {
+                final Object value = ctx.context.attr(attrKey);
+                if (value != null) {
+                    return value;
```

Comment:
Should we raise an exception if `value ` is not an instance of `resolver.elementType()`?
In this context, we could provide a detailed error message instead of  `ClassCastException`.

Reply:
@ikhoon nim, thanks for your comment! 
I agree with you. 

As you said, I added some codes to produce detail comment for developer! 
Please review again 🙇‍♂️ 
```
...
                final Object value = ctx.context.attr(attrKey);
                final Class<?> elementType = resolver.elementType().isPrimitive() ?
                                             Primitives.wrap(resolver.elementType()) :
                                             resolver.elementType();
                if (!elementType.isInstance(value)) {
                    throw new ClassCastException("The value " + value +
                                                 " cannot be cast to elementyType " +
                                                 elementType.getSimpleName());
                }
                if (value != null) {
                    return value;
                }
...
```

Reply:
@ikhoon nim, Sorry to say, IMHO, it is difficult to implement it.  😅 
In case of `value` has not container, it is very easy. however, `value` has container type, it is very complex.

For example, 
- AttributeKey : `AttributeKey<List<String>>` 
- AttributeValue : `ImmutableList.of("A");`
- Parameter declared `@Attribute('...') List<String> value`
in this time, `resolver.containerType()` is `class.java.util.ArrayList`, however, value type is `com.google.common.collect.SingletonImmutableList`. 

Thus, when the `value` is of a `Container Type`, checking for casting can be complex, which may actually lead to more unexpected errors at runtime. 

What do you think?

Reply:
https://github.com/line/armeria/blob/e132c5da5627392a807e75c191ea8833b0c40e36/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java#L1326-L1330

when `AnnotatedValueResolver` are builded, `AnnotatedValueResolver#getContainerType()` is called. 
That time, `ArrayList.class` are returned. 

However, `ArrayList.class` cannot be cast to `ImmutableList.class`.  😢 

> Should we raise an exception if value is not an instance of resolver.elementType()?
> In this context, we could provide a detailed error message instead of ClassCastException.

IMHO, big change will follow to reach your comment.
Do you have any good idea? 🙇‍♂️

Reply:
I missed that point. We may add `rawContainerType` to `AnnotatedValueResolver` to know the original type before normalized.

Reply:
@ikhoon nim, thanks for your guideline.
i added `rawContainerType` and `getRawContainerType()` simply and can achieve goal finally!
Also, i added test cases to test `ClassCastException`.

When you have free time, would you please check it again?  🙇‍♂️